symbolic-symcache/src/new/transform.rs

```diff
@@ -0,0 +1,98 @@
+//! Utilities that transform the Data to be written to a SymCache.
+
+use std::borrow::Cow;
+
+use symbolic_debuginfo::macho::BcSymbolMap;
+
+/// A Function record to be written to the SymCache.
+#[non_exhaustive]
+pub struct Function<'s> {
+    /// The functions name.
+    pub name: Cow<'s, str>,
+    /// The compilation directory of the function.
+    pub comp_dir: Option<Cow<'s, str>>,
+}
+
+/// A File to be written to the SymCache.
+#[non_exhaustive]
+pub struct File<'s> {
+    /// The file name.
+    pub name: Cow<'s, str>,
+    /// The optional directory prefix.
+    pub directory: Option<Cow<'s, str>>,
+    /// The optional compilation directory prefix.
+    pub comp_dir: Option<Cow<'s, str>>,
+}
+
+/// A Source Location (File + Line) to be written to the SymCache.
+#[non_exhaustive]
+pub struct SourceLocation<'s> {
+    /// The [`File`] part of this [`SourceLocation`].
+    pub file: File<'s>,
+    /// The line number.
+    pub line: u32,
+}
+
+/// A transformer that is applied to each [`Function`] and [`SourceLocation`] record in the SymCache.
+pub trait Transformer {
+    /// Transforms a [`Function`] record.
+    ///
+    /// This can be used for example to de-obfuscate a functions name.
+    fn transform_function<'f>(&'f self, f: Function<'f>) -> Function<'f> {
+        f
+    }
+
+    /// Transforms a [`SourceLocation`].
+    ///
+    /// This can be used for example to apply a Source Mapping in case an intermediate compilation
+    /// step might have introduced an indirection, or to de-obfuscate the [`File`] information.
+    fn transform_source_location<'f>(&'f self, sl: SourceLocation<'f>) -> SourceLocation<'f> {
+        sl
+    }
+}
+
+// This is essentially just a newtype in order to implement `Debug`.
+#[derive(Default)]
+pub(crate) struct Transformers(pub Vec<Box<dyn Transformer>>);
```

Comment:
Using `&dyn Transformer` would avoid the boxing, but unfortunately infect everything with a lifetime, and we can’t add a lifetime to the outer `SymCacheConverter` backwards compatibly.